### ISSUE TITLE
Support selectors in environment.yml

### ISSUE THREAD

**Issue by cdeil** [conda/conda#8089]
We are shipping conda environments to users via `environment.yml` files like this (see [docs](https://docs.gammapy.org/0.9/getting-started.html#install):
```
curl -O https://gammapy.org/download/install/gammapy-0.9-environment.yml
conda env create -f gammapy-0.9-environment.yml
conda activate gammapy-0.9
```

Now there are two optional dependencies that don't support Windows.

It would be great if we could use selectors as in conda build `meta.yml` also in `environment.yml` to declare that these dependencies should simply be not installed on Windows, like this:
```
sherpa # [not win]
healpy # [not win]
```

I found https://stackoverflow.com/questions/32869042 - that's someone else asking for the same feature. Using `conda build` to configure this doesn't really work for us, we are building one package with many optional dependencies, and then want to provide several environments that are more minimal or full-featured to users.

**Comment by soapy1** [comment]
related to https://github.com/conda/conda/pull/7258

**Comment by SylvainCorlay** [comment]
:+1:  on this. I would love environments to support preprocessing selectors.

**Comment by jjhelmus** [comment]
This would be a great addition to conda.  If anyone in the community is interested in work on this the help would be most welcome and appreciated.

**Comment by GollyJer** [comment]
Yup. Ran into this today as well. 
Wish I had time to contribute. 😒

**Comment by silvanmelchior** [comment]
I had the same problem (packages depending on platform and whether there is a GPU available) and wrote a tiny parser which transfers _environment.yml.meta_ into _environment.yml_ files. For your use-case, you could do

```
sherpa [platform != win32]
healpy [platform != win32]
```

It's of course not the same as if it was supported native, because you need to either add the parser as submodule to your repo or just copy it over, but maybe you find it useful. The repo is

https://github.com/silvanmelchior/cme_parser

**Comment by silvanmelchior** [comment]
`cmd_parser` does not need to be installed, it's a single python file with no dependencies which could be run in the default conda env (exactly because of this chicken and the egg problem). But I agree, you still need to download it somehow and execute it, which indeed is a bit hacky in your use-case.

Anyways, I copy, if someone would code a PR, this would be smoother.

**Comment by GollyJer** [comment]
@silvanmelchior `cme_parser` is cool. Thanks!

**Comment by chris-rands** [comment]
Great idea! Is conda planning to support this? Did anyone submit a PR?

**Comment by analog-cbarber** [comment]
I think that to do this right, you would want to refactor the existing selector code in https://github.com/conda/conda-build/blob/master/conda_build/metadata.py and move it to some common location in the conda repo (not sure where) so there isn't two copies of the selector code and so that external tools can make use of selector filtering.

**Comment by FaustinCarter** [comment]
> I think that to do this right, you would want to refactor the existing selector code in https://github.com/conda/conda-build/blob/master/conda_build/metadata.py and move it to some common location in the conda repo (not sure where) so there isn't two copies of the selector code and so that external tools can make use of selector filtering.

This is a really _really_ good idea. The conda/constructor project is already copying over that parsing functionality from conda-build and would strongly benefit from having a centralized location to import that functionality from that doesn't require importing all of conda-build.

**Comment by rijobro** [comment]
It's a shame, our current solution to only install on linux or to not install on windows is to call pip from conda which defeats the purpose of conda:

```yaml
dependencies:
  - numpy>=1.17
  - pip
  - pip:
    - pytype; platform_system != "Windows"
    - cucim; platform_system == "Linux"
 ```

**Comment by JosueCom** [comment]
Has there been any updates to integrating this functionality? It would also be beneficial if one can check to see if a GPU is available in the system

**Comment by bryanloz-xilinx** [comment]
Had a shower thought today about optimizing our separate Linux/Windows conda flows by using such a feature…

Looks like I’m not the first.

**Comment by jaimergp** [comment]
If this is going to be implemented, please please let's move away from semantic comments like `# [win]` and let's instead use something parsable. [The new recipe.yaml format uses minijinja if-then-else clauses for this](https://github.com/conda-incubator/ceps/blob/main/cep-13.md#selectors) and while not as compact, it is easier to manipulate by the tooling that will have to build on that format.

**Comment by johndpope** [comment]
if you want a populate library to implement on this 8,000 star repo would be good one - there's like a sprawl of diffferent requirements.txt - out of control.  https://github.com/bmaltais/kohya_ss

**Comment by travishathaway** [comment]
Just going to chime in here by saying that I think the answer is developing a new format for `environment.yaml` and not tacking on to it with such features.

It would be great for us to have something similar to `pixi.toml` (`conda.toml`?). This feature would come in the form of a separate plugin though and not be integrated into conda itself.

**Comment by analog-cbarber** [comment]
A new environment file format would be fine as well, but only if it happens!

The thing is, the conda ecosystem ALREADY supports these selectors foir build and has for years, so it is not really a bizarre idea to support it for regular environment files as well.

**Comment by jamesmyatt** [comment]
Try conda-lock or pixi if you need this. Raw conda is not really designed for multi-platform environments.

**Comment by analog-cbarber** [comment]
But projects that use conda ARE designed for multi-platform environments and need a way to manage dependency difference among those environments. Since conda does not provide any easy way to do that, everyone is forced to come up with their own hack to get past this.

**Comment by jamesmyatt** [comment]
Yes. That's one of the reasons why tools like [conda-lock](https://github.com/conda/conda-lock) and [pixi](https://pixi.sh) exist as part of the conda ecosystem. No need to invent your own hacks anymore.

Can you explain why you can't use conda-lock or pixi for your multi-platform project? Unless you can do that, the conda developers will never prioritise this issue.

**Comment by analog-cbarber** [comment]
Pixi looks interesting, but that seems like it is pretty much a replacement for conda on the client side.

I will definitely consider using that, but the point of this issue is a much simpler solution to an obvious problem. Sometimes people just want a single environment file that will work for more than one platform. I think it is pretty obvious that switching to an entirely different way of managing packages is a much higher bar for users than simply continuing to use the conda tools.

Regardless, I developed my own workaround for this a long time ago, so I don't really have a personal stake in it at this point.

**Comment by jaimergp** [comment]
See https://github.com/conda/ceps/pull/81 for context.

It is unlikely that `conda` will implement `# [selectors]` (semantic comments are a bad idea), but a new revision of the environment.yml format should end up providing something that satisfies this need.

**Comment by jamesmyatt** [comment]
If you want a "single environment file that will work for more than one platform", then use conda-lock.

But the conda ecosystem is moving on from semantic comments (for example: [cep-0013](https://github.com/conda/ceps/blob/main/cep-0013.md) and (https://github.com/conda/ceps/blob/main/cep-0014.md)), so I don't see any of the developers spending time on it.

**Comment by jjerphan** [comment]
For your information, [selectors are supported by mamba as specified by the CEP linked by Jaime above](https://github.com/conda/ceps/pull/81/files#diff-bebfa3f1d70547b3695124927ba31659f76574696cb2ae65c971578b2069ff5fR168-R179).

In short, the syntax to use in environments' specifications is:
```yaml
dependencies:
 - sel(linux): xtensor
 - sel(osx): make
 - sel(unix): cmake
 - sel(win): ninja
```

**Comment by jamesmyatt** [comment]
That's interesting. Looks it's mamba-specific syntax currently (e.g. https://github.com/mamba-org/mamba/issues/1788) and I don't see any mention in the mamba docs though. Hopefully it will get implemented in conda-lock at least also.
